bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/CommerceRedirectServlet.java

```diff
@@ -0,0 +1,138 @@
+import com.adobe.cq.commerce.magento.graphql.ProductInterfaceQueryDefinition;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+import com.day.cq.wcm.api.PageManagerFactory;
+
+@Component(
+    service = Servlet.class,
+    immediate = true,
+    property = {
+        ServletResolverConstants.SLING_SERVLET_METHODS + "=" + HttpConstants.METHOD_GET,
+        ServletResolverConstants.SLING_SERVLET_RESOURCE_TYPES + "=" + CommerceRedirectServlet.RESOURCE_TYPE,
+        ServletResolverConstants.SLING_SERVLET_SELECTORS + "=" + CommerceRedirectServlet.SELECTOR,
+        ServletResolverConstants.SLING_SERVLET_EXTENSIONS + "=" + CommerceRedirectServlet.EXTENSION
+    })
+@Designate(ocd = UrlProviderConfiguration.class)
+public class CommerceRedirectServlet extends SlingSafeMethodsServlet {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CommerceRedirectServlet.class);
+
+    protected static final String SELECTOR = "cifredirect";
+    protected static final String EXTENSION = "html";
+    protected static final String RESOURCE_TYPE = "core/cif/components/structure/page/v1/page";
+
+    private Pair<IdentifierLocation, ProductIdentifierType> productIdentifierConfig;
+    private Page productPage;
+
+    @Reference
+    private UrlProvider urlProvider;
+
+    @Reference
+    PageManagerFactory pageManagerFactory;
+
+    @Activate
+    public void activate(UrlProviderConfiguration conf) {
+        productIdentifierConfig = Pair.of(conf.productIdentifierLocation(), conf.productIdentifierType());
+    }
+
+    @Override
+    protected void doGet(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
+        String suffix = request.getRequestPathInfo().getSuffix();
+
+        if (suffix == null) {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Missing redirect suffix.");
+            return;
+        }
+
+        String[] suffixInfo = suffix.substring(1).split("/");
+        if (suffixInfo.length != 2) {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Redirect suffix has wrong format.");
+            return;
+        }
+
+        PageManager pageManager = pageManagerFactory.getPageManager(request.getResourceResolver());
+
+        Page currentPage = pageManager.getContainingPage(request.getResource());
+        productPage = SiteNavigation.getProductPage(currentPage);
+
+        if ("product".equals(suffixInfo[0])) {
+            redirectToProductPage(request, response, suffixInfo[1]);
+        } else {
+            response.sendError(HttpServletResponse.SC_NOT_IMPLEMENTED, "The requested redirect is not available.");
+        }
+    }
+
+    private void redirectToProductPage(SlingHttpServletRequest request, SlingHttpServletResponse response, String sku) throws IOException {
+        ParamsBuilder params = new ParamsBuilder();
+
+        if (productIdentifierConfig.getRight().equals(ProductIdentifierType.SKU)) {
```

Comment:
This logic will not be needed any more with CIF-2062 implemented. All knowledge about URLProvider internals and config will be encapsulated. I suggest we pause this or rebased on top of `issues/CIF-2062` branch.